src/AddImageChecks.cpp

```diff
@@ -13,15 +13,50 @@ using std::pair;
 using std::string;
 using std::vector;
 
+namespace {
+
+// Sotres buffer asserts Stmt's ordered by buffer name and assert type.
+class BufferAsserts {
+public:
+    enum Type { DimsNoOverflow,
+                Constrained,
+                Proposed,
+                Required,
+                ElemSize,
+                HostAllignment,
+                HostNonNull,
+    };
+    void insert(const BufferAsserts::Type &type, const string &buffer_name, const Stmt &stmt) {
+        asserts_data_[buffer_name][type].push_back(stmt);
+    }
+
+    // Iterates over the asserts of a certain types and run apply functor to each Stmt.
+    void process_types(std::function<void(const Stmt &)> apply, const std::vector<Type> &types) {
+        for (const auto &current_buffer_asserts : asserts_data_) {
+            for (const Type type : types) {
+                const auto &asserts = current_buffer_asserts.second.find(type);
+                if (asserts == current_buffer_asserts.second.end()) continue;
+                for (const Stmt &stmt : asserts->second) {
+                    apply(stmt);
+                }
+            }
+        }
+    }
+
+private:
+    std::map<string, std::map<Type, std::vector<Stmt>>> asserts_data_;
+};
+}  // namespace
+
 /* Find all the externally referenced buffers in a stmt */
 class FindBuffers : public IRGraphVisitor {
 public:
     struct Result {
         Buffer<> image;
         Parameter param;
         Type type;
-        int dimensions{0};
-        bool used_on_host{false};
+        int dimensions{ 0 };
```

Comment:
Don't make unrelated formatting changes (here and elsewhere); those are based saved for separate PRs, as it makes the "real" changes much harder to parse out in the diff.